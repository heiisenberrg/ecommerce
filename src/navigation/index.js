import * as React from 'react';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import Cart from '../cart';
import Home from '../home';

const Main = createNativeStackNavigator();

export function MainStack(props) {
  return (
    <Main.Navigator
      gesturesEnabled={false}
      screenOptions={{headerShown: false}}>
      <Main.Screen name={'Home'} component={Home} headerMode="none" />
      <Main.Screen name={'Cart'} component={Cart} headerMode="none" />
    </Main.Navigator>
  );
}
