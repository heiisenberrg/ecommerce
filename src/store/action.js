import { GET_PRODUCT_LIST, SET_PRODUCTS, SET_SELECTED_PRODUCT } from './actionTypes';

export const getProducts = () => ({
    type: GET_PRODUCT_LIST
});

export const setProducts = (data) => ({
    type: SET_PRODUCTS,
    data
});

export const setSelectedProducts = (data) => ({
    type: SET_SELECTED_PRODUCT,
    data
});