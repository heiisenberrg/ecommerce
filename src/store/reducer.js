import {
  SET_PRODUCTS,
  SET_SELECTED_PRODUCT,
  GET_PRODUCT_LIST,
} from './actionTypes';

const initialState = {
  loading: false,
  products: [],
  selectedProduct: {},
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_PRODUCT_LIST:
      return {
        ...state,
        loading: true,
      };
    case SET_PRODUCTS:
      return {
        ...state,
        products: action.data,
        loading: false,
      };
    case SET_SELECTED_PRODUCT:
      return {
        ...state,
        selectedProduct: action.data,
      };
    default:
      return state;
  }
};
