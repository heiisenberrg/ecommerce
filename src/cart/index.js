import React from 'react';
import {
  View,
  Text,
  FlatList,
  Image,
  TouchableOpacity,
  StyleSheet,
} from 'react-native';
import {connect} from 'react-redux';
import {setSelectedProducts} from '../store/action';

const Cart = ({selectedProduct, navigation, setSelectedProducts}) => {
  const incrementItem = id => {
    let selectedItems = {...selectedProduct};
    selectedItems[id].count += 1;
    setSelectedProducts(selectedItems);
  };

  const decrementItem = id => {
    let selectedItems = {...selectedProduct};
    if (selectedItems[id].count > 1) {
      selectedItems[id].count -= 1;
    } else {
      delete selectedItems[id];
    }
    setSelectedProducts(selectedItems);
  };

  const calculateTotalPrice = () => {
    let total = 0;
    Object.keys(selectedProduct).map(id => {
      total += selectedProduct[id].price * selectedProduct[id].count;
    });
    return total.toFixed(2);
  };

  const renderItem = ({item: id, index}) => {
    const item = selectedProduct[id];
    return (
      <View key={`${item.id}-${index}`} style={styles.itemContainer}>
        <View style={styles.imgContainer}>
          <View style={styles.imgContent}>
            <Image source={{url: item.img}} style={styles.img} />
          </View>
          <View style={styles.descContainer}>
            <Text numberOfLines={4} style={styles.productNameText}>
              {item.name}
            </Text>
            <Text style={styles.priceText}>${item.price}</Text>
          </View>
        </View>
        <View style={styles.countContainer}>
          <TouchableOpacity
            activeOpacity={0.7}
            style={styles.decrementContainer}
            onPress={() => decrementItem(id)}>
            <Text style={styles.countText}>-</Text>
          </TouchableOpacity>
          <View style={styles.countContent}>
            <Text style={styles.countText}>{item.count}</Text>
          </View>
          <TouchableOpacity
            activeOpacity={0.7}
            style={styles.incrementContainer}
            onPress={() => incrementItem(id)}>
            <Text style={styles.countText}>+</Text>
          </TouchableOpacity>
        </View>
      </View>
    );
  };

  return (
    <View style={styles.container}>
      <View style={styles.headerContainer}>
        <TouchableOpacity
          activeOpacity={0.7}
          style={styles.backContent}
          onPress={() => navigation.goBack()}>
          <Text style={styles.backText}>{'<'}</Text>
        </TouchableOpacity>
        <Text style={styles.headerTitle}>Cart</Text>
      </View>
      <FlatList
        data={Object.keys(selectedProduct)}
        contentContainerStyle={styles.containerStyle}
        extraData={selectedProduct}
        keyExtractor={(_, index) => index}
        renderItem={renderItem}
      />
      <View style={styles.totalContent}>
        <Text
          style={
            styles.totalText
          }>{`Total\t\t$ ${calculateTotalPrice()}`}</Text>
      </View>
    </View>
  );
};

const mapStateToProps = state => ({
  selectedProduct: state.selectedProduct,
});

const mapDispatchToProps = dispatch => ({
  setSelectedProducts: data => dispatch(setSelectedProducts(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Cart);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#e0e0e0',
  },
  headerContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 52,
    marginLeft: 16,
  },
  backContent: {
    backgroundColor: '#fff',
    width: 30,
    height: 30,
    borderRadius: 8,
    justifyContent: 'center',
  },
  backText: {
    color: '#000',
    textAlign: 'center',
  },
  headerTitle: {
    fontSize: 32,
    color: '#000',
    fontWeight: '900',
    textAlign: 'center',
    marginLeft: 24,
  },
  containerStyle: {
    flex: 1,
    marginHorizontal: 16,
    marginVertical: 24,
  },
  totalContent: {
    position: 'absolute',
    width: '100%',
    bottom: 0,
    paddingBottom: 60,
    backgroundColor: 'rgba(0, 0, 0, 0.4)',
    padding: 12,
  },
  totalText: {
    fontSize: 24,
    fontWeight: '700',
    color: '#fff',
    textAlign: 'right',
    marginRight: 12,
  },
  itemContainer: {
    marginBottom: 16,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  imgContainer: {
    flex: 0.7,
    flexDirection: 'row',
  },
  imgContent: {
    width: 100,
    height: 140,
    backgroundColor: '#fff',
    overflow: 'hidden',
    borderRadius: 8,
  },
  img: {
    width: 100,
    height: 140,
    resizeMode: 'stretch',
  },
  descContainer: {
    marginLeft: 12,
    paddingTop: 16,
  },
  productNameText: {
    fontSize: 14,
    fontWeight: '600',
    color: '#000',
    marginBottom: 8,
  },
  priceText: {
    fontSize: 24,
    color: '#000',
    fontWeight: 'bold',
  },
  countContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  decrementContainer: {
    backgroundColor: '#d0d0d0',
    width: 28,
    height: 32,
    borderBottomLeftRadius: 4,
    borderTopLeftRadius: 4,
    justifyContent: 'center',
  },
  countText: {
    color: '#000',
    textAlign: 'center',
  },
  countContent: {
    width: 36,
    height: 32,
    backgroundColor: '#fff',
    justifyContent: 'center',
  },
  incrementContainer: {
    backgroundColor: '#d0d0d0',
    width: 28,
    height: 32,
    borderTopRightRadius: 4,
    borderBottomRightRadius: 4,
    justifyContent: 'center',
  },
});
