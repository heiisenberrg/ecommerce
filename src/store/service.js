import axios from 'axios';
import {setProducts, getProducts} from './action';

export const getProductList = () => async (dispatch) => {
  try {
    dispatch(getProducts());
    const result = await axios({
      method: 'GET',
      url: 'https://my-json-server.typicode.com/benirvingplt/products/products',
    });
    dispatch(setProducts(result.data));
  } catch (err) {
    throw new Error(err);
  }
};
