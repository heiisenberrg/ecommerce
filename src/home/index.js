import React, {useEffect} from 'react';
import {
  View,
  Text,
  FlatList,
  ActivityIndicator,
  Image,
  TouchableOpacity,
  StyleSheet,
} from 'react-native';
import {connect} from 'react-redux';
import {setSelectedProducts} from '../store/action';
import {getProductList} from '../store/service';

const Home = ({
  loading,
  products,
  selectedProduct,
  getProducts,
  setSelectedProducts,
  navigation,
}) => {
  useEffect(() => {
    getProducts();
  }, []);

  const onProductPress = item => {
    let selectedItems = {...selectedProduct};
    if (selectedProduct[item.id]) {
      delete selectedItems[item.id];
    } else {
      selectedItems = {
        ...selectedItems,
        [item.id]: {
          ...item,
          count: 1,
        },
      };
    }
    setSelectedProducts(selectedItems);
  };

  const renderItem = ({item, index}) => {
    return (
      <View key={`${item.id}-${index}`} style={styles.productContainer}>
        <Image source={{url: item.img}} style={styles.productImg} />
        <View style={styles.descriptionContainer}>
          <Text numberOfLines={2} style={styles.productNameText}>
            {item.name}
          </Text>
          <View style={styles.priceContainer}>
            <Text style={styles.priceText}>${item.price}</Text>
            <TouchableOpacity
              activeOpacity={0.7}
              style={styles.addToCartContainer}
              onPress={() => onProductPress(item)}>
              <Text style={styles.addToCartText}>
                {selectedProduct[item.id] ? 'Remove' : 'Add to Cart'}
              </Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>
    );
  };

  return (
    <View style={styles.container}>
      <Text style={styles.headerTitle}>Products</Text>
      {loading ? (
        <View style={styles.loadingContainer}>
          <ActivityIndicator animating={true} size={'large'} color={'#000'} />
        </View>
      ) : (
        <FlatList
          data={products}
          contentContainerStyle={styles.containerStyle}
          numColumns={2}
          extraData={products}
          keyExtractor={item => item.id}
          renderItem={renderItem}
        />
      )}
      {Object.keys(selectedProduct).length > 0 && (
        <TouchableOpacity
          activeOpacity={0.7}
          style={styles.checkoutContainer}
          onPress={() => navigation.navigate('Cart')}>
          <Text style={styles.checkoutText}>CHECKOUT</Text>
        </TouchableOpacity>
      )}
    </View>
  );
};

const mapStateToProps = state => ({
  loading: state.loading,
  products: state.products,
  selectedProduct: state.selectedProduct,
});

const mapDispatchToProps = dispatch => ({
  getProducts: () => dispatch(getProductList()),
  setSelectedProducts: data => dispatch(setSelectedProducts(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#e0e0e0',
  },
  headerTitle: {
    fontSize: 32,
    color: '#000',
    fontWeight: '900',
    textAlign: 'center',
    marginTop: 52,
  },
  loadingContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  containerStyle: {
    flex: 1,
    marginHorizontal: 12,
    marginVertical: 24,
  },
  checkoutContainer: {
    position: 'absolute',
    bottom: 48,
    right: 24,
    backgroundColor: '#fff',
    paddingHorizontal: 16,
    paddingVertical: 12,
    borderRadius: 8,
  },
  checkoutText: {
    fontSize: 20,
    fontWeight: '700',
    color: '#000',
  },
  productContainer: {
    width: '48%',
    backgroundColor: '#fff',
    marginHorizontal: 4,
    borderRadius: 8,
    marginBottom: 16,
    overflow: 'hidden',
  },
  productImg: {
    width: '100%',
    height: 300,
    resizeMode: 'stretch',
  },
  descriptionContainer: {
    position: 'absolute',
    width: '100%',
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.4)',
    padding: 12,
  },
  productNameText: {
    fontSize: 14,
    fontWeight: '600',
    color: '#fff',
    marginBottom: 8,
  },
  priceContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  priceText: {
    fontSize: 24,
    color: '#fff',
    fontWeight: 'bold',
  },
  addToCartContainer: {
    backgroundColor: '#fff',
    paddingHorizontal: 6,
    paddingVertical: 8,
    borderRadius: 4,
  },
  addToCartText: {
    fontSize: 12,
    fontWeight: '700',
    color: '#000',
  },
});
