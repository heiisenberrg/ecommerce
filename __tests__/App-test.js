import * as React from 'react';
import {Provider} from 'react-redux';
import {NavigationContainer} from '@react-navigation/native';
import {render, screen, fireEvent} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import {store} from '../src/store';
import {MainStack} from '../src/navigation';
import Home from '../src/home';
import Cart from '../src/cart';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

describe('render screen correctly', () => {
  test('Home', () => {
    const component = (
      <Provider store={store}>
        <NavigationContainer>
          <Home />
        </NavigationContainer>
      </Provider>
    );
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Cart', () => {
    const component = (
      <Provider store={store}>
        <NavigationContainer>
          <Cart />
        </NavigationContainer>
      </Provider>
    );
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Testing navigation between screens', () => {
  test('page contains the header and 4 items', async () => {
    const component = (
      <Provider store={store}>
        <NavigationContainer>
          <MainStack />
        </NavigationContainer>
      </Provider>
    );

    render(component);

    const header = await screen.findByText('Products');
    const items = await screen.findAllByText(/Add to Cart/);

    expect(header).toBeTruthy();
    expect(items.length).toBe(4);
  });

  test('clicking on one item to the cart and checkout to cart screen', async () => {
    const component = (
      <Provider store={store}>
        <NavigationContainer>
          <MainStack />
        </NavigationContainer>
      </Provider>
    );

    render(component);
    const toClick = await screen.findAllByText(/Add to Cart/);

    fireEvent(toClick[0], 'press');
    const checkoutClick = await screen.findByText('CHECKOUT');

    fireEvent(checkoutClick, 'press');
    const newHeader = await screen.findByText('Cart');
    const newBody = await screen.findByText(/Total/);

    expect(newHeader).toBeTruthy();
    expect(newBody).toBeTruthy();
  });
});
