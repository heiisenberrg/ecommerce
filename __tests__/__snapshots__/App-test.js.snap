// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render screen correctly Cart 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#e0e0e0",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginLeft": 16,
        "marginTop": 52,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "height": 30,
          "justifyContent": "center",
          "opacity": 1,
          "width": 30,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "textAlign": "center",
          }
        }
      >
        &lt;
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 32,
          "fontWeight": "900",
          "marginLeft": 24,
          "textAlign": "center",
        }
      }
    >
      Cart
    </Text>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "marginHorizontal": 16,
        "marginVertical": 24,
      }
    }
    data={Array []}
    extraData={Object {}}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.4)",
        "bottom": 0,
        "padding": 12,
        "paddingBottom": 60,
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 24,
          "fontWeight": "700",
          "marginRight": 12,
          "textAlign": "right",
        }
      }
    >
      Total		$ 0.00
    </Text>
  </View>
</View>
`;

exports[`render screen correctly Home 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#e0e0e0",
      "flex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#000",
        "fontSize": 32,
        "fontWeight": "900",
        "marginTop": 52,
        "textAlign": "center",
      }
    }
  >
    Products
  </Text>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "marginHorizontal": 12,
        "marginVertical": 24,
      }
    }
    data={Array []}
    extraData={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
